server/webhook.go

```diff
@@ -280,81 +289,101 @@ func (s *Server) processPushEvent(ctx context.Context, repositoryList []*api.Rep
 	return createdMessages, nil
 }
 
+type fileType int
+
+const (
+	unknownFileType fileType = iota
+	migrationFileType
+	schemaFileType
+)
+
+func getFileInfo(fileItem vcs.DistinctFileItem, repositoryList []*api.Repository) (*db.MigrationInfo, fileType, *api.Repository, error) {
+	var migrationInfo *db.MigrationInfo
+	var fType fileType
+	var fileRepositoryList []*api.Repository
+	for _, repository := range repositoryList {
+		if !strings.HasPrefix(fileItem.FileName, repository.BaseDirectory) {
+			log.Debug("Ignored file outside the base directory",
+				zap.String("file", fileItem.FileName),
+				zap.String("base_directory", repository.BaseDirectory),
+			)
+			continue
+		}
+
+		// NOTE: We do not want to use filepath.Join here because we always need "/" as the path separator.
+		mi, err := db.ParseMigrationInfo(fileItem.FileName, path.Join(repository.BaseDirectory, repository.FilePathTemplate))
+		if err != nil {
+			log.Error("Failed to parse migration file info",
+				zap.Int("project", repository.ProjectID),
+				zap.String("file", fileItem.FileName),
+				zap.Error(err),
+			)
+			continue
+		}
+		if mi != nil {
+			migrationInfo = mi
+			fType = migrationFileType
+			fileRepositoryList = append(fileRepositoryList, repository)
+			continue
+		}
+
+		si, err := db.ParseSchemaFileInfo(repository.BaseDirectory, repository.SchemaPathTemplate, fileItem.FileName)
+		if err != nil {
+			log.Debug("Failed to parse schema file info",
+				zap.String("file", fileItem.FileName),
+				zap.Error(err),
+			)
+			continue
+		}
+		if si != nil {
+			migrationInfo = si
+			fType = schemaFileType
+			fileRepositoryList = append(fileRepositoryList, repository)
+			continue
+		}
+	}
+
+	switch len(fileRepositoryList) {
+	case 0:
+		return nil, unknownFileType, nil, errors.Errorf("file change is not associated with any project")
+	case 1:
+		return migrationInfo, fType, fileRepositoryList[0], nil
+	default:
+		var projectList []string
+		for _, repository := range fileRepositoryList {
+			projectList = append(projectList, repository.Project.Name)
+		}
+		return nil, unknownFileType, nil, errors.Errorf("file change should be associated with exactly one project but found %s", strings.Join(projectList, ","))
```

Comment:
if this is an error, should the caller pass repository instead of the repository list?

Reply:
This function attempts to match the file with a repository in the list. This is the case when a file matches multiple repositories (projects) in a mono-repo setup.